core/src/executor/evaluate/mod.rs

```diff
@@ -294,6 +301,75 @@ async fn evaluate_function<'a, 'b: 'a, 'c: 'a, T: GStore>(
             let exprs = stream::iter(exprs).then(eval).try_collect().await?;
             f::concat(exprs)
         }
+        Function::Custom { name, exprs } => {
+            let custom_func = storage
+                .fetch_function(name)
+                .await?
+                .ok_or_else(|| EvaluateError::UnsupportedFunction(name.to_string()))?;
+            let args = stream::iter(exprs)
+                .then(eval)
+                .try_collect::<Vec<_>>()
+                .await?;
+            let args = args
+                .into_iter()
+                .map(Value::try_from)
+                .collect::<Result<Vec<_>>>()?;
+
+            let fargs = &custom_func.args;
+
+            let dargs = fargs
+                .iter()
+                .filter_map(|y| y.default.as_ref())
+                .collect::<Vec<_>>();
+            let dargs = stream::iter(dargs)
+                .then(eval)
+                .try_collect::<Vec<_>>()
+                .await?;
+            let dargs = dargs
+                .into_iter()
+                .map(Value::try_from)
+                .collect::<Result<Vec<_>>>()?;
+            let mut dargs = dargs.iter();
+
+            let min = fargs.len() - dargs.len();
+            let max = fargs.len();
+
+            let value = if (min..=max).contains(&args.len()) {
+                let mut hm = StdHashMap::new();
+
+                fargs
+                    .iter()
+                    .enumerate()
+                    .try_for_each(|(i, farg)| -> Result<()> {
+                        let arg = args.get(i).unwrap_or(&Value::Null);
+                        arg.validate_type(&farg.data_type)?;
+                        arg.validate_null(farg.default.is_some())?;
+                        let value = if arg.is_null() {
+                            dargs.next().unwrap()
+                        } else {
+                            arg
+                        };
+                        hm.insert(farg.name.to_owned(), value.to_owned());
+                        Ok(())
+                    })?;
+
+                let row = Row::Map(hm);
+                let rowcontext = RowContext::new(name, Cow::Owned(row), None);
+                let context = Rc::new(rowcontext);
+
+                eval_with_context(&custom_func.body, context).await
+            } else {
+                Err((EvaluateError::FunctionArgsLengthNotWithinRange {
+                    name: custom_func.func_name.to_owned(),
+                    expected_minimum: min,
+                    expected_maximum: max,
+                    found: args.len(),
+                })
+                .into())
+            };
+
+            Ok(value?)
```

Comment:
We can simplify the codes.
I added some comments for your information. when you use this, you can remove the comments.
Behaviors are mostly same but there's slight difference in handling default expr.

```rust
let CustomFunction {
    func_name,
    args,
    body,
} = storage
    .fetch_function(name)
    .await?
    .ok_or_else(|| EvaluateError::UnsupportedFunction(name.to_string()))?;

// it would make easy to read code by splitting
// validation and evaluation.
let min = args.iter().filter(|arg| arg.default.is_none()).count();
let max = args.len();
if !(min..=max).contains(&exprs.len()) {
    return Err((EvaluateError::FunctionArgsLengthNotWithinRange {
        name: func_name.to_owned(),
        expected_minimum: min,
        expected_maximum: max,
        found: exprs.len(),
    })
    .into());
}

let exprs = exprs.iter().chain(
    args.iter()
        .skip(max - exprs.len())
        .filter_map(|arg| arg.default.as_ref()),
);

let context = stream::iter(args.iter().zip(exprs))
    .then(|(arg, expr)| async move {
        // if user provides NULL explicitly,
        // then NULL should be used, not replaced by default expr.
        eval(expr)
            .await?
            .try_into_value(&arg.data_type, true) // Evaluated::try_into_value
            .map(|value| (arg.name.to_owned(), value))
    })
    .try_collect()
    .await
    .map(|row| {
        let row = Cow::Owned(Row::Map(row));
        let context = RowContext::new(name, row, None);

        Some(Rc::new(context))
    })?;

// function body should not capture outer context so
// aggregated also needs to be `None`
evaluate(storage, context, None, body).await
```